dwave/optimization/src/simplex.cpp

```diff
@@ -0,0 +1,563 @@
+            }
+        }
+    }
+    return status;
+}
+
+Matrix construct_T(std::span<const double> c, double c0, const Matrix& A,
+                   std::span<const double> b) {
+    Matrix T(A.n() + 2, A.m() + A.n() + 1);
+    // Copy in A to T[:n, :m] and b.T to A[:, -1]
+    for (ssize_t i = 0; i < A.n(); i++) {
+        double sign = b[i] < 0 ? -1.0 : 1.0;
+        T(i, -1) = sign * b[i];
+        for (ssize_t j = 0; j < A.m(); j++) {
+            T(i, j) = sign * A(i, j);
+        }
+    }
+
+    // T[:, m:m+n] = I
+    for (ssize_t i = 0; i < A.n(); i++) {
+        T(i, A.m() + i) = 1;
+    }
+
+    // Row objective
+    for (ssize_t i = 0; i < A.m(); i++) {
+        T(A.n(), i) = c[i];
+    }
+    T(A.n(), T.m() - 1) = c0;
+
+    // Row pseudo objective
+    for (ssize_t j = 0; j < A.m(); j++) {
+        for (ssize_t i = 0; i < A.n(); i++) {
+            T(A.n() + 1, j) -= T(i, j);
+        }
+    }
+    for (ssize_t i = 0; i < A.n(); i++) {
+        T(A.n() + 1, T.m() - 1) -= T(i, T.m() - 1);
+    }
+
+    return T;
+}
+
+SolveResult _linprog_simplex(std::span<const double> c, double c0, const Matrix& A,
+                             std::span<const double> b, ssize_t max_iterations, double tolerance,
+                             bool bland) {
+    assert(static_cast<ssize_t>(c.size()) == A.m());
+
+    std::vector<ssize_t> basis;
+    for (ssize_t i = 0; i < A.n(); i++) {
+        basis.push_back(i + A.m());
+    }
+
+    Matrix T = construct_T(c, c0, A, b);
+
+    ssize_t phase = 1;
+    SolveResult status =
+            _solve_simplex(T, A.n(), basis, max_iterations, tolerance, phase, bland, 0);
+
+    if (std::abs(T(-1, -1)) < tolerance) {
+        Matrix newT(A.n() + 1, A.m() + 1);
+
+        // newT[:, :m] = T[:-1, :m]
+        for (ssize_t i = 0; i < newT.n(); i++) {
+            for (ssize_t j = 0; j < A.m(); j++) {
+                newT(i, j) = T(i, j);
+            }
+        }
+        // newT[:, -1] = T[:, -1]
+        for (ssize_t i = 0; i < newT.n(); i++) {
+            newT(i, -1) = T(i, -1);
+        }
+
+        std::swap(T, newT);
+    } else {
+        status.solve_status = SolveResult::SolveStatus::FAILURE_NO_FEASIBLE_START;
+    }
+
+    if (status.solve_status == SolveResult::SolveStatus::SUCCESS) {
+        phase = 2;
+        assert(T.n() == A.n() + 1 && T.m() == A.m() + 1);
+        SolveResult status2 = _solve_simplex(T, A.n(), basis, max_iterations, tolerance, phase,
+                                             bland, status.num_iterations);
+
+        std::swap(status, status2);
+    }
+
+    SolveResult result(status);
+
+    std::vector<double> solution(A.m(), 0.0);
+    for (ssize_t i = 0; i < A.n(); i++) {
+        if (basis[i] < A.m()) {
+            solution[basis[i]] = T(i, -1);
+        }
+    }
+    result.set_partial_solution(std::move(solution));
+
+    return result;
+}
+
+struct LP {
+    LP(std::vector<double>&& c, double c0, Matrix&& A, std::vector<double>&& b)
```

Comment:
Learning opportunity for me: why pass in the 'reference to a reference' here?  I'm not following how that works with, what looks to me like, the vector's copy-constructor...

Reply:
This is an [Rvalue reference](https://en.cppreference.com/w/cpp/language/reference). Basically it's saying that it must take ownership of the vector.